### ISSUE TITLE
gcc builds: u64 to f64 casts are broken

### ISSUE THREAD

**Issue by Dragorn421** [zeldaret/oot#1527]
(at least on the compiler I use
```
$ mips-linux-gnu-gcc --version
mips-linux-gnu-gcc (Ubuntu 12.3.0-1ubuntu1~23.04) 12.3.0
```
)

Consider this code casting a u64 to a f64:

```diff
diff --git a/src/boot/idle.c b/src/boot/idle.c
index 30c244876..93a5f3495 100644
--- a/src/boot/idle.c
+++ b/src/boot/idle.c
@@ -35,6 +35,12 @@ void Main_ThreadEntry(void* arg) {
 }
 
 void Idle_ThreadEntry(void* arg) {
+
+    {
+        static u64 value = 421;
+        osSyncPrintf("Look at my value = %f\n", (f64)value);
+    }
+
     osSyncPrintf("アイドルスレッド(idleproc)実行開始\n");
     osSyncPrintf("作製者    : %s\n", gBuildTeam);
     osSyncPrintf("作成日時  : %s\n", gBuildDate);
```

It should print something like "`Look at my value = 421.0`"
But it prints:
> `Look at my value = 0.000000`

asm:
```mips
80000650 <Idle_ThreadEntry>:
80000650:       27bdffd8        addiu   sp,sp,-40
80000654:       3c058001        lui     a1,0x8001
80000658:       8ca581b4        lw      a1,-32332(a1)
8000065c:       afb00020        sw      s0,32(sp)
80000660:       00808025        move    s0,a0
80000664:       3c048001        lui     a0,0x8001
80000668:       8c8481b0        lw      a0,-32336(a0)
8000066c:       afbf0024        sw      ra,36(sp)
80000670:       0c001ff7        jal     80007fdc <__floatundidf>
80000674:       00000000        nop
80000678:       f7a00018        sdc1    $f0,24(sp)
8000067c:       8fa7001c        lw      a3,28(sp)
80000680:       3c048001        lui     a0,0x8001
80000684:       24849de8        addiu   a0,a0,-25112
80000688:       0c000798        jal     80001e60 <osSyncPrintf>
8000068c:       8fa60018        lw      a2,24(sp)
```

for some reason the compiler uses `__floatundidf` and even seems to call it assuming it takes a u64 (a0 and a1 are filled), but `__floatundidf` is for u32 (== unsigned long in this case) to f64

---------

gcc bug or bad compile flags?

**Comment by Dragorn421** [comment]
Should be fixed by https://github.com/zeldaret/oot/pull/1529 (thanks tharo :muscle: )
Needs testing, leaving open until testing confirms it's fixed

**Comment by Dragorn421** [comment]
Fixed by #1529

Testing details/results:

```diff
diff --git a/src/boot/idle.c b/src/boot/idle.c
index 30c244876..e7ab4a546 100644
--- a/src/boot/idle.c
+++ b/src/boot/idle.c
@@ -19,6 +19,74 @@ f32 gViConfigYScale = 1.0;
 void Main_ThreadEntry(void* arg) {
     OSTime time;
 
+    {
+        static s8 value_s8 = 42;
+        static s16 value_s16 = 421;
+        static s32 value_s32 = 421;
+        static s64 value_s64 = 421;
+
+        static s8 nvalue_s8 = -42;
+        static s16 nvalue_s16 = -421;
+        static s32 nvalue_s32 = -421;
+        static s64 nvalue_s64 = -421;
+
+        static u8 value_u8 = 42;
+        static u16 value_u16 = 421;
+        static u32 value_u32 = 421;
+        static u64 value_u64 = 421;
+
+        static f32 value_f32 = 421.0f;
+        static f64 value_f64 = 421.0;
+        static f32 nvalue_f32 = -421.0f;
+        static f64 nvalue_f64 = -421.0;
+
+        osSyncPrintf("value_s8  (int)(f32)=%d (f64)=%f\n", (int)(f32)value_s8, (f64)value_s8);
+        osSyncPrintf("value_s16 (int)(f32)=%d (f64)=%f\n", (int)(f32)value_s16, (f64)value_s16);
+        osSyncPrintf("value_s32 (int)(f32)=%d (f64)=%f\n", (int)(f32)value_s32, (f64)value_s32);
+        osSyncPrintf("value_s64 (int)(f32)=%d (f64)=%f\n", (int)(f32)value_s64, (f64)value_s64);
+
+        osSyncPrintf("nvalue_s8  (int)(f32)=%d (f64)=%f\n", (int)(f32)nvalue_s8, (f64)nvalue_s8);
+        osSyncPrintf("nvalue_s16 (int)(f32)=%d (f64)=%f\n", (int)(f32)nvalue_s16, (f64)nvalue_s16);
+        osSyncPrintf("nvalue_s32 (int)(f32)=%d (f64)=%f\n", (int)(f32)nvalue_s32, (f64)nvalue_s32);
+        osSyncPrintf("nvalue_s64 (int)(f32)=%d (f64)=%f\n", (int)(f32)nvalue_s64, (f64)nvalue_s64);
+
+        osSyncPrintf("value_u8  (int)(f32)=%d (f64)=%f\n", (int)(f32)value_u8, (f64)value_u8);
+        osSyncPrintf("value_u16 (int)(f32)=%d (f64)=%f\n", (int)(f32)value_u16, (f64)value_u16);
+        osSyncPrintf("value_u32 (int)(f32)=%d (f64)=%f\n", (int)(f32)value_u32, (f64)value_u32);
+        osSyncPrintf("value_u64 (int)(f32)=%d (f64)=%f\n", (int)(f32)value_u64, (f64)value_u64);
+
+        osSyncPrintf("value_f32 (int)(s16)=%d\n", (int)(s16)value_f32);
+        osSyncPrintf("value_f32 (int)(s32)=%d\n", (int)(s32)value_f32);
+        osSyncPrintf("value_f32 (int)(s64)=%d\n", (int)(s64)value_f32);
+        osSyncPrintf("value_f32 (int)(u16)=%d\n", (int)(u16)value_f32);
+        osSyncPrintf("value_f32 (int)(u32)=%d\n", (int)(u32)value_f32);
+        osSyncPrintf("value_f32 (int)(u64)=%d\n", (int)(u64)value_f32);
+
+        osSyncPrintf("value_f64 (int)(s16)=%d\n", (int)(s16)value_f64);
+        osSyncPrintf("value_f64 (int)(s32)=%d\n", (int)(s32)value_f64);
+        osSyncPrintf("value_f64 (int)(s64)=%d\n", (int)(s64)value_f64);
+        osSyncPrintf("value_f64 (int)(u16)=%d\n", (int)(u16)value_f64);
+        osSyncPrintf("value_f64 (int)(u32)=%d\n", (int)(u32)value_f64);
+        osSyncPrintf("value_f64 (int)(u64)=%d\n", (int)(u64)value_f64);
+
+        osSyncPrintf("nvalue_f32 (int)(s16)=%d\n", (int)(s16)nvalue_f32);
+        osSyncPrintf("nvalue_f32 (int)(s32)=%d\n", (int)(s32)nvalue_f32);
+        osSyncPrintf("nvalue_f32 (int)(s64)=%d\n", (int)(s64)nvalue_f32);
+        osSyncPrintf("nvalue_f32 (int)(u16)=%d\n", (int)(u16)nvalue_f32);
+        osSyncPrintf("nvalue_f32 (int)(u32)=%d\n", (int)(u32)nvalue_f32);
+        osSyncPrintf("nvalue_f32 (int)(u64)=%d\n", (int)(u64)nvalue_f32);
+
+        osSyncPrintf("nvalue_f64 (int)(s16)=%d\n", (int)(s16)nvalue_f64);
+        osSyncPrintf("nvalue_f64 (int)(s32)=%d\n", (int)(s32)nvalue_f64);
+        osSyncPrintf("nvalue_f64 (int)(s64)=%d\n", (int)(s64)nvalue_f64);
+        osSyncPrintf("nvalue_f64 (int)(u16)=%d\n", (int)(u16)nvalue_f64);
+        osSyncPrintf("nvalue_f64 (int)(u32)=%d\n", (int)(u32)nvalue_f64);
+        osSyncPrintf("nvalue_f64 (int)(u64)=%d\n", (int)(u64)nvalue_f64);
+    }
+
+    while (1)
+        ;
+
     osSyncPrintf("mainx 実行開始\n");
     DmaMgr_Init();
     osSyncPrintf("codeセグメントロード中...");
```

For reference, with ido:
```
value_s8  (int)(f32)=42 (f64)=42.000000
value_s16 (int)(f32)=421 (f64)=421.000000
value_s32 (int)(f32)=421 (f64)=421.000000
<emulated cpu (in ares) seems to crash here,
after investigating,
both `(int)(f32)value_s64` and `(f64)value_s64` trip it up
so huh idk>
```

With gcc and bad revision 8913c4fa0d2cec19db6e8b4d8598125977243c5d^ == dadd1f84934f49cfc8fe881f8c49ead6e48b8b76 (had the issue):
```
value_s8  (int)(f32)=42 (f64)=42.000000
value_s16 (int)(f32)=421 (f64)=421.000000
value_s32 (int)(f32)=421 (f64)=421.000000
value_s64 (int)(f32)=0 (f64)=0.000000  # bad
nvalue_s8  (int)(f32)=-42 (f64)=-42.000000
nvalue_s16 (int)(f32)=-421 (f64)=-421.000000
nvalue_s32 (int)(f32)=-421 (f64)=-421.000000
nvalue_s64 (int)(f32)=-1 (f64)=-1.000000  # bad
value_u8  (int)(f32)=42 (f64)=42.000000
value_u16 (int)(f32)=421 (f64)=421.000000
value_u32 (int)(f32)=421 (f64)=421.000000
value_u64 (int)(f32)=0 (f64)=0.000000  # bad
value_f32 (int)(s16)=421
value_f32 (int)(s32)=421
value_f32 (int)(s64)=25  # bad
value_f32 (int)(u16)=421
value_f32 (int)(u32)=421
value_f32 (int)(u64)=25  # bad
value_f64 (int)(s16)=421
value_f64 (int)(s32)=421
value_f64 (int)(s64)=25  # bad
value_f64 (int)(u16)=421
value_f64 (int)(u32)=421
value_f64 (int)(u64)=25  # bad
nvalue_f32 (int)(s16)=-421
nvalue_f32 (int)(s32)=-421
nvalue_f32 (int)(s64)=27  # bad
nvalue_f32 (int)(u16)=65115  # weird
nvalue_f32 (int)(u32)=-421  # weird
nvalue_f32 (int)(u64)=27  # very weird
nvalue_f64 (int)(s16)=-421
nvalue_f64 (int)(s32)=-421
nvalue_f64 (int)(s64)=27  # bad
nvalue_f64 (int)(u16)=65115  # weird
nvalue_f64 (int)(u32)=-421  # weird
nvalue_f64 (int)(u64)=27  # very weird
```
"bad" is wrong, it's most of the stuff involving u64, s64 or f64
"weird" is ok, it's negative float/double to unsigned integers, which afaik is undefined/implementation-defined

With gcc and main (042a5c553020404a3e49834614327917c34334a8):
```
value_s8  (int)(f32)=42 (f64)=42.000000
value_s16 (int)(f32)=421 (f64)=421.000000
value_s32 (int)(f32)=421 (f64)=421.000000
value_s64 (int)(f32)=421 (f64)=421.000000
nvalue_s8  (int)(f32)=-42 (f64)=-42.000000
nvalue_s16 (int)(f32)=-421 (f64)=-421.000000
nvalue_s32 (int)(f32)=-421 (f64)=-421.000000
nvalue_s64 (int)(f32)=-421 (f64)=-421.000000
value_u8  (int)(f32)=42 (f64)=42.000000
value_u16 (int)(f32)=421 (f64)=421.000000
value_u32 (int)(f32)=421 (f64)=421.000000
value_u64 (int)(f32)=421 (f64)=421.000000
value_f32 (int)(s16)=421
value_f32 (int)(s32)=421
value_f32 (int)(s64)=421
value_f32 (int)(u16)=421
value_f32 (int)(u32)=421
value_f32 (int)(u64)=421
value_f64 (int)(s16)=421
value_f64 (int)(s32)=421
value_f64 (int)(s64)=421
value_f64 (int)(u16)=421
value_f64 (int)(u32)=421
value_f64 (int)(u64)=421
nvalue_f32 (int)(s16)=-421
nvalue_f32 (int)(s32)=-421
nvalue_f32 (int)(s64)=-421
nvalue_f32 (int)(u16)=65115  # weird
nvalue_f32 (int)(u32)=-421  # weird
nvalue_f32 (int)(u64)=0  # weird
nvalue_f64 (int)(s16)=-421
nvalue_f64 (int)(s32)=-421
nvalue_f64 (int)(s64)=-421
nvalue_f64 (int)(u16)=65115  # weird
nvalue_f64 (int)(u32)=-421  # weird
nvalue_f64 (int)(u64)=0  # weird
```
everything's ok (again negative float/double to unsigned integers are weird but ok)

**Comment by Dragorn421** [comment]
For completeness, turns out ido choked on the test code because I put it in boot and some float/int conversion intrinsics used by ido are in code
The test code in `Main`, running after code is loaded, works, printing:
```
value_s8  (int)(f32)=42 (f64)=42.000000
value_s16 (int)(f32)=421 (f64)=421.000000
value_s32 (int)(f32)=421 (f64)=421.000000
value_s64 (int)(f32)=421 (f64)=421.000000
nvalue_s8  (int)(f32)=-42 (f64)=-42.000000
nvalue_s16 (int)(f32)=-421 (f64)=-421.000000
nvalue_s32 (int)(f32)=-421 (f64)=-421.000000
nvalue_s64 (int)(f32)=-421 (f64)=-421.000000
value_u8  (int)(f32)=42 (f64)=42.000000
value_u16 (int)(f32)=421 (f64)=421.000000
value_u32 (int)(f32)=421 (f64)=421.000000
value_u64 (int)(f32)=421 (f64)=421.000000
value_f32 (int)(s16)=421
value_f32 (int)(s32)=421
value_f32 (int)(s64)=421
value_f32 (int)(u16)=421
value_f32 (int)(u32)=421
value_f32 (int)(u64)=421
value_f64 (int)(s16)=421
value_f64 (int)(s32)=421
value_f64 (int)(s64)=421
value_f64 (int)(u16)=421
value_f64 (int)(u32)=421
value_f64 (int)(u64)=421
nvalue_f32 (int)(s16)=-421
nvalue_f32 (int)(s32)=-421
nvalue_f32 (int)(s64)=-421
nvalue_f32 (int)(u16)=65535  # weird
nvalue_f32 (int)(u32)=-1  # weird
nvalue_f32 (int)(u64)=-1  # weird
nvalue_f64 (int)(s16)=-421
nvalue_f64 (int)(s32)=-421
nvalue_f64 (int)(s64)=-421
nvalue_f64 (int)(u16)=65535  # weird
nvalue_f64 (int)(u32)=-1  # weird
nvalue_f64 (int)(u64)=-1  # weird
```

which is ok, again the weird bits with negative to uint conversions, interestingly it behaves different
